expression/builtin_time.go

```diff
@@ -2039,69 +2041,102 @@ type timestampFunctionClass struct {
+	evalTps, argLen := []evalTp{tpString}, len(args)
+	if argLen == 2 {
+		evalTps = append(evalTps, tpString)
+	}
+	fsp := c.getDefaultFsp4Type(args[0].GetType())
+    if argLen == 2 {
+		fsp = mathutil.Max(fsp, c.getDefaultFsp4Type(args[1].GetType()))
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpDatetime, evalTps...)
+    bf.tp.Decimal, bf.tp.Flen = fsp, 19
+	if fsp != 0 {
+		bf.tp.Flen += 1 + fsp
+	}
+
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var sig builtinFunc
+	if argLen == 2 {
+		sig = &builtinTimestamp2ArgsSig{baseTimeBuiltinFunc{bf}}
+	} else {
+		sig = &builtinTimestamp1ArgSig{baseTimeBuiltinFunc{bf}}
+	}
 	return sig.setSelf(sig), nil
 }
 
-type builtinTimestampSig struct {
-	baseBuiltinFunc
+type builtinTimestamp1ArgSig struct {
+	baseTimeBuiltinFunc
 }
 
-// eval evals a builtinTimestampSig.
+// evalTime evals a builtinTimestamp1ArgSig.
 // See https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_timestamp
-func (b *builtinTimestampSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
-	}
-	if args[0].IsNull() {
-		return
+func (b *builtinTimestamp1ArgSig) evalTime(row []types.Datum) (types.Time, bool, error) {
+	s, isNull, err := b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return types.Time{}, isNull, errors.Trace(err)
 	}
-	var arg0 types.Time
-	switch tp := args[0].Kind(); tp {
-	case types.KindInt64, types.KindUint64:
-		arg0, err = types.ParseDatetimeFromNum(args[0].GetInt64())
-		if err != nil {
-			return d, errors.Trace(err)
-		}
-	case types.KindString, types.KindBytes, types.KindMysqlDecimal, types.KindFloat32, types.KindFloat64:
-		s, err1 := args[0].ToString()
-		if err1 != nil {
-			return d, errors.Trace(err1)
-		}
-		arg0, err = types.ParseTime(s, mysql.TypeDatetime, getFsp(s))
-		if err != nil {
-			return d, errors.Trace(err)
-		}
-	case types.KindMysqlTime:
-		arg0 = args[0].GetMysqlTime()
-	default:
-		return d, errors.Errorf("Unknown args type for timestamp %d", tp)
+	tm, err := types.ParseTime(s, mysql.TypeDatetime, getFsp(s))
+	if err != nil {
+		return types.Time{}, true, errors.Trace(handleInvalidTimeError(b.ctx, err))
 	}
-	if len(args) == 1 {
-		d.SetMysqlTime(arg0)
-		return
+	return tm, false, nil
+}
+
+type builtinTimestamp2ArgsSig struct {
+	baseTimeBuiltinFunc
+}
+
+// evalTime evals a builtinTimestamp2ArgsSig.
+// See https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_timestamp
+func (b *builtinTimestamp2ArgsSig) evalTime(row []types.Datum) (types.Time, bool, error) {
+	ctx := b.ctx.GetSessionVars().StmtCtx
+	arg0, isNull, err := b.args[0].EvalString(row, ctx)
+	if isNull || err != nil {
+		return types.Time{}, isNull, errors.Trace(err)
 	}
-	// If exists args[1].
-	s, err := args[1].ToString()
+	tm, err := types.ParseTime(arg0, mysql.TypeDatetime, getFsp(arg0))
```

Comment:
how about add a cast and let child 0 return time, so we can call `b.args[0].EvalTime(row, ctx)`

Reply:
The inserted `CAST` uses an default value 6, which is not as expected. examples are as follows:
**In mysql:**
```sql
mysql> select timestamp('2017-08-24 00:00:00.01');
Field   1:  `timestamp('2017-08-24 00:00:00.01')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     22
Max_length: 22
Decimals:   2
Flags:      BINARY 


+-------------------------------------+
| timestamp('2017-08-24 00:00:00.01') |
+-------------------------------------+
| 2017-08-24 00:00:00.01              |
+-------------------------------------+
1 row in set (0.00 sec)

mysql> select timestamp('2017-08-24 00:00:00.1234');
Field   1:  `timestamp('2017-08-24 00:00:00.1234')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     24
Max_length: 24
Decimals:   4
Flags:      BINARY 


+---------------------------------------+
| timestamp('2017-08-24 00:00:00.1234') |
+---------------------------------------+
| 2017-08-24 00:00:00.1234              |
+---------------------------------------+
1 row in set (0.00 sec)

```

**In TiDB, using a `CAST`:**
```sql
mysql> select timestamp('2017-08-24 00:00:00.01');
Field   1:  `timestamp('2017-08-24 00:00:00.01')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     22
Max_length: 22
Decimals:   2
Flags:      BINARY 


+-------------------------------------+
| timestamp('2017-08-24 00:00:00.01') |
+-------------------------------------+
| 2017-08-24 00:00:00.010000              |
+-------------------------------------+
1 row in set (0.00 sec)

mysql> select timestamp('2017-08-24 00:00:00.1234');
Field   1:  `timestamp('2017-08-24 00:00:00.1234')`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     24
Max_length: 24
Decimals:   4
Flags:      BINARY 


+---------------------------------------+
| timestamp('2017-08-24 00:00:00.1234') |
+---------------------------------------+
| 2017-08-24 00:00:00.123400              |
+---------------------------------------+
1 row in set (0.00 sec)